packages/gatsby/index.d.ts

```diff
@@ -687,9 +697,66 @@ export interface SetFieldsOnGraphQLNodeTypeArgs extends ParentSpanPluginArgs {
   traceId: "initial-setFieldsOnGraphQLNodeType"
 }
 
-export interface SourceNodesArgs extends ParentSpanPluginArgs {
+export interface GatsbyGraphQLObjectType<TSource = any, TContext = any> {
+  kind: "OBJECT"
+  config: ComposeObjectTypeConfig<TSource, TContext>
+}
+
+interface GatsbyGraphQLInputObjectType {
+  kind: "INPUT_OBJECT"
+  config: ComposeInputObjectTypeConfig
+}
+
+interface GatsbyGraphQLUnionType<TSource = any, TContext = any> {
+  kind: "UNION"
+  config: ComposeUnionTypeConfig<TSource, TContext>
+}
+
+interface GatsbyGraphQLInterfaceType<TSource = any, TContext = any> {
+  kind: "INTERFACE"
+  config: ComposeInterfaceTypeConfig<TSource, TContext>
+}
+
+interface GatsbyGraphQLEnumType {
+  kind: "ENUM"
+  config: ComposeEnumTypeConfig
+}
+
+interface GatsbyGraphQLScalarType {
+  kind: "SCALAR"
+  config: ComposeScalarTypeConfig
+}
+
+export type GatsbyGraphQLType<TSource = any, TContext = any> =
+  | GatsbyGraphQLObjectType
+  | GatsbyGraphQLInputObjectType
+  | GatsbyGraphQLUnionType<TSource, TContext>
+  | GatsbyGraphQLInterfaceType<TSource, TContext>
+  | GatsbyGraphQLEnumType
+  | GatsbyGraphQLScalarType
+
+export interface SourceNodesSchemaObject<TSource = any, TContext = any> {
+  buildObjectType(
+    config: ComposeObjectTypeConfig<TSource, TContext>
+  ): GatsbyGraphQLObjectType<TSource, TContext>
+  buildUnionType(
+    config: ComposeUnionTypeConfig<TSource, TContext>
+  ): GatsbyGraphQLUnionType<TSource, TContext>
+  buildInterfaceType(
+    config: ComposeInterfaceTypeConfig<TSource, TContext>
+  ): GatsbyGraphQLInterfaceType<TSource, TContext>
+  buildInputObjectType(
+    config: ComposeInputObjectTypeConfig
+  ): GatsbyGraphQLInputObjectType
+  buildEnumType(config: ComposeEnumTypeConfig): GatsbyGraphQLEnumType
+  buildScalarType(config: ComposeScalarTypeConfig): GatsbyGraphQLScalarType
+}
+
+export interface SourceNodesArgs<TSource = any, TContext = any>
+  extends ParentSpanPluginArgs {
   traceId: "initial-sourceNodes"
   waitForCascadingActions: boolean
+  schema: SourceNodesSchemaObject<TSource, TContext>
```

Comment:
Thanks, I'll update 👍

Reply:
One side effect is that the generic types are now consumed by all the interfaces and methods that inherit / use `NodePluginArgs`.

Reply:
This will be correct, because we do really pass `schema` to each gatsby-node hook. Here's list of common things passed to those functions:

https://github.com/gatsbyjs/gatsby/blob/2de96171ec57b1d9d17fb983162e224c14a9e9fa/packages/gatsby/src/utils/api-runner-node.js#L174-L205

Might not be the most practical thing, given that `createTypes` can only be used in limited amount of APIs: https://github.com/gatsbyjs/gatsby/blob/2de96171ec57b1d9d17fb983162e224c14a9e9fa/packages/gatsby/src/redux/actions/restricted.js#L443-L446

But it will be at least correct ;)